### PR TITLE
batches: implement commit signing via GitHub apps

### PR DESCRIPTION
**Pull request by BolajiOlajide**
Closes https://github.com/sourcegraph/sourcegraph/issues/48611.

## Test plan

Reminder to tag security-team before merging this into `main`.

### CODE REVIEW

File: enterprise/cmd/frontend/internal/auth/githubappauth/middleware.go
Comment (courier-new) [inline enterprise/cmd/frontend/internal/auth/githubappauth/middleware.go:1]: Here we're passing a couple more parameters back and forth between the frontend, the middleware, and GitHub and store a couple of these in the redis cache as well. We collect a bit more information on the new GitHub App installation, and we handle errors a bit more gracefully.

File: enterprise/cmd/worker/internal/githubapps/worker/installation_backfill.go
Comment (courier-new) [inline enterprise/cmd/worker/internal/githubapps/worker/installation_backfill.go:1]: This background worker requests the current set of installations for a GitHub App on GitHub and then syncs our records in the DB to match it. We also opportunistically perform this sync when an admin visits the GitHub App page in the web UI.

File: enterprise/internal/batches/reconciler/executor.go
Comment (courier-new) [inline enterprise/internal/batches/reconciler/executor.go:1]: Here we now detect if we're pushing commits to a GitHub code host and try to authenticate with a GitHub App installation. If we can, we'll push a commit with that authentication in order to sign it. If we can't, we just return.

**Review by courier-new — COMMENTED** [review]
Hey @sourcegraph/security-code-review would love a quick review of a couple things here! This PR enables commit signing with Batch Changes via GitHub Apps. We were already able to connect GitHub Apps to Sourcegraph prior to this, so this PR just builds off of that a bit. As such, there's not a whole lot that I think warrants your attention, but I called out just a few areas of interest in the hopes you could give it some eyes! Notably:

- The middleware layer we use to facilitate the GitHub App connection process, keep state, and handle the installation flow redirects
- A new background worker used for recording and syncing additional metadata about individual GitHub App Installations
- In the Batch Changes code, authenticating and pushing commits via a GitHub App Installation, if we find one that works for a given repo

**Review by evict — CHANGES_REQUESTED** [review]
Looks good! I only have one recommendation for the caching bit. 👍

**Review by evict — APPROVED** [review]

**Pull request merged**